lib/Dialect/Torch/Transforms/RecomposeComplexOps.cpp

```diff
@@ -9,15 +9,47 @@
 
 #include "PassDetail.h"
 
+#include "mlir/IR/Matchers.h"
 #include "mlir/Transforms/GreedyPatternRewriteDriver.h"
 #include "torch-mlir/Dialect/Torch/IR/TorchOps.h"
+#include "torch-mlir/Dialect/Torch/IR/TorchTypes.h"
 #include "torch-mlir/Dialect/Torch/Transforms/Passes.h"
 #include "torch-mlir/Dialect/Torch/Utils/Utils.h"
+#include <climits>
+#include <cstdint>
 
 using namespace mlir;
 using namespace mlir::torch;
 using namespace mlir::torch::Torch;
 
+// TODO Comparing directly to INT64_MAX/INT64_MIN seems fragile.
+// This is a potential general way of implementing the clamping in terms of
+// other torch ops. https://github.com/llvm/torch-mlir/pull/2005
+//
+// def to_valid_dim(dim, max_dim):
+//  dim = torch.ops.prim.min(dim, max_dim)
+//  dim = torch.ops.prim.max(dim, -max_dim)
+//  is_neg = torch.ops.aten.lt(dim, 0)
+//  is_neg_int = torch.ops.aten.Int.bool(is_neg)
+//  return (dim + max_dim) * is_neg_int + dim * (1 - is_neg_int)
+//
+// dim_size = torch.ops.aten.size.int(slice, slice.get_dim())
+// start = to_valid_dim(slice.get_start(), dim_size)
+// end = to_valid_dim(slice.get_end(), dim_size)
+//
+// Expect this to change to the above once op support is there.
+// The signature will also likely change.
+static Value clampDimToValidRange(OpBuilder &b, Location loc, Value clampVal,
```

Comment:
I think
```
static Value clampDimToValidRange(OpBuilder &b, Location loc, Value clampVal,
                                  int64_t clampInt, Value tensor, Value dim) {
  Value dimSize = b.create<AtenSizeIntOp>(loc, tensor, dim);
  if (clampInt < 0) {
    return b.create<AtenAddIntOp>(loc, dimSize, clampVal);
  } else  {
    Value dimIsSmaller = b.create<AtenLtIntOp>(loc, dimSize, clampVal);
      auto dimIsSmallerInt = b.create<AtenIntBoolOp>(loc, dimIsSmaller);

      Value dimMasked = b.create<AtenMulIntOp>(loc, dimIsSmallerInt, dimSize);
      auto minusOne =b.create<ConstantIntOp>(loc, b.getI64IntegerAttr(-1));
      auto notDimIsSmallerInt = b.create<AtenAddIntOp>(loc, dimIsSmallerInt, minusOne);
      Value endMasked = b.create<AtenMulIntOp>(loc, notDimIsSmallerInt, clampVal);

      return b.create<AtenAddIntOp>(
          loc, dimMasked, endMasked);
  }
  return clampVal;
}
```
will work. I wrote it when I was debugging an issue with another backend where the specified `end` was greater than the tensor size.

Reply:
@gpetters94

Reply:
That's causing this error:

```
python: /home/gap/work/mlir-npcomp/externals/llvm-project/llvm/../mlir/include/mlir/IR/StorageUniquerSupport.h:174: static ConcreteT mlir::detail::StorageUserBase<mlir::torch::Torch::ValueTensorType, mlir::torch::Torch::BaseTensorType, mlir::torch::Torch::detail::ValueTensorTypeStorage, mlir::detail::TypeUniquer>::get(mlir::MLIRContext *, Args...) [ConcreteT = mlir::torch::Torch::ValueTensorType, BaseT = mlir::torch::Torch::BaseTensorType, StorageT = mlir::torch::Torch::detail::ValueTensorTypeStorage, UniquerT = mlir::detail::TypeUniquer, Traits = <>, Args = <std::optional<llvm::ArrayRef<long>>, mlir::Type>]: Assertion `succeeded(ConcreteT::verify(getDefaultDiagnosticEmitFn(ctx), args...))' failed.
```

If I remember correctly that was the same kind of pure Torch op implementation Ramiro recommended caused. I'll dig through and see if I can find the log from then, but if it's the same then this solution will also require a lot of work to make viable.